com.unity.multiplayer.mlapi/Runtime/AOI/ClientObjectMap.cs

```diff
@@ -0,0 +1,110 @@
+// Although this is (currently) inside the MLAPI package, it is intentionally
+//  totally decoupled from MLAPI with the intention of allowing it to live
+//  in its own package
+namespace MLAPI.AOI
+{
+    // To establish a Client Object Map, instantiate a ClientObjMapNodeBase, then
+    //  add more nodes to it (and those nodes) as desired
+   public class ClientObjMapNode<CLIENT, OBJECT>
+   {
+        // set this delegate if you want a function called when
+        //  object 'obj' is being de-spawned
+        public delegate void DespawnDelegate(OBJECT obj);
+        public DespawnDelegate OnDespawn;
+
+        // to dynamically compute objects to be added each 'QueryFor' call,
+        //  assign this delegate to your handler
+        public delegate void QueryDelegate(CLIENT c, HashSet<OBJECT> results);
+        public QueryDelegate OnQuery;
+
+        public delegate void BypassDelegate(HashSet<OBJECT> results);
+        public BypassDelegate OnBypass;
+
+        public ClientObjMapNode() : base()
+        {
+            ChildNodes = new List<ClientObjMapNode<CLIENT, OBJECT>>();
```

Comment:
I think you're referring to C/C++ typedef aliases. There is no such thing in C# and I'd even prefer full expanded version of templated classes. In C++ world, it makes more sense since they're combining too many std:: stuff (even for some primitives) but in C#, I really want to be more explicit and expanded — and these kind of more than 2/3 templated combos are rare in C# compared to C/C++ world :)